From 05cd15aa7f72c597755238e5a61145bff0421c90 Mon Sep 17 00:00:00 2001
From: Hansong Zhang <hsz@google.com>
Date: Thu, 12 Jul 2018 10:51:30 -0700
Subject: [PATCH 3/7] DO NOT MERGE Fix OOB read in process_l2cap_cmd

Test: manual
Bug: 79488381
Change-Id: I723866ed40d3647fed99875f659bb95df96a6969
(cherry picked from commit 54c6a9dfd52ac6711d6f2101d233b276b2e3bb53)
CVE-2018-9484
---
 stack/l2cap/l2c_main.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/stack/l2cap/l2c_main.c b/stack/l2cap/l2c_main.c
index 75ae855..577e1b3 100644
--- a/stack/l2cap/l2c_main.c
+++ b/stack/l2cap/l2c_main.c
@@ -582,6 +582,10 @@ static void process_l2cap_cmd (tL2C_LCB *p_lcb, UINT8 *p, UINT16 pkt_len)
                     /* sanity check option length */
                     if ((cfg_len + L2CAP_CFG_OPTION_OVERHEAD) <= cmd_len)
                     {
+                        if (p + cfg_len > p_next_cmd) {
+                          android_errorWriteLog(0x534e4554, "79488381");
+                          return;
+                        }
                         p += cfg_len;
                         if ((cfg_code & 0x80) == 0)
                         {
-- 
2.7.4

